Add --watch flag to pr checks command

endorama opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

Hello maintainers! 

I'm using GitHub for a lot of projects, and found myself in the same situation as explained in https://github.com/cli/cli/issues/3669

CI checks takes a long time to run, and manually refreshing the GitHub web interface is suboptimal.

Following the discussion in the linked issue, I added a `--wait` flag to `pr checks` command. When this flag is used, the command does not exit immediately but report on current check statuses and wait until the number of pending checks is 0.

[![asciicast](https://asciinema.org/a/442125.svg)](https://asciinema.org/a/442125)

Fixes #3669

Implementation wise this PR refactor the current `pr checks` code to extract the data gathering logic and splits the 3 "modes" (web mode, wait mode, single execution mode) to reuse the current functionalities in multiple places. Existing functionalities have not been modified.  

I did not work on the `--watch` as I think it requires a bit more discussion around the desired UI, while the `--wait` is useful to everyone having to wait for a slow CI with a simpler UI.  
With this command is possible to execute use cases like "merge after all checks have been successful" from the GitHub CLI.

Happy to receive feedback about this, there are possible improvements but I wanted to keep this focused on a single change.  
Thank you!

endorama committed:
pr-checks: extract webMode

endorama committed:
pr-checks: extract checks information collection

endorama committed:
pr-checks: extract output utilities

samcoe commented:
@endorama Thanks for the contribution! After discussing with the team offline we have come to the consensus that it would be confusing to have both a `--wait` and `--watch` flag for this command. We think having a `--watch` flag with a UI that refreshes in the same manner as `gh run watch` is the direction we want to go. From a quick glance at the PR it does look like many of the changes you made could be reused to pivot in this new direction, is that something you would be interested in working on? If not, can close this PR and allow other contributors to work on the issue?

endorama commented:
Hi @samcoe, thank you for looking into this and sorry for missing the notification. I'll work on the `--watch` flag.

endorama committed:
pr-checks: implement watch flag

endorama commented:
@samcoe I removed the `--wait` and added the `--watch` flag, copying the behaviour from `run watch` command.

There are 2 main differences:
- I used  a different ANSI escape code (`\033[H\033[2J` instead of `\x1b[2J`); this code properly clear the screen and reset position to 0,0 (at least on Linux but I can't run it on other platforms at the moment)
- I added `SIGINT` handling to exit with code `1` (still checks pending) or `0` (all checks completed; this is an edge case)

Let me know and thanks for the patience.

samcoe commented:
@endorama Thanks for the continued effort here. Two quick comments. 

> I used a different ANSI escape code (\033[H\033[2J instead of \x1b[2J); this code properly clear the screen and reset position to 0,0 (at least on Linux but I can't run it on other platforms at the moment)

The current code `\x1b[2J` is not working for you on Linux? If it is working I would suggest we continue to use it. If it isn't working we should change it in both places. 
cc @vilmibm Since you wrote `run watch` which uses that screen clearing code.

> I added SIGINT handling to exit with code 1 (still checks pending) or 0 (all checks completed; this is an edge case)

Seems to me that this mostly adds unnecessary complexity for a no real improvement. If the user manually cancels the command then they should get the SIGINT exit code.

I still got to do a full code review but just wanted to address those two changes you brought up.

endorama commented:
> The current code \x1b[2J is not working for you on Linux? 

Is only partially working: this code only clears the screen without resetting the cursor to 0,0 position. 
This is visually strange, as the text is attached to the bottom of the terminal window, suggesting to scroll. I'll wait on some hints from @vilmibm about this.

> Seems to me that this mostly adds unnecessary complexity for a no real improvement. If the user manually cancels the command then they should get the SIGINT exit code.

I'll remove it.

endorama committed:
pr-checks: remove SIGINT interceptor

endorama committed:
pr-checks: exit with error if some task has failed

samcoe commented on pkg/cmd/pr/checks/checks.go:
```diff
@@ -14,6 +13,8 @@ import (
 	"github.com/spf13/cobra"
 )
 
+const defaultInterval int = 30
```
I think it probably would be better to drop this down to 5 or 10 seconds for the default. `run watch` uses 3 seconds but maybe checks should refresh at a slower rate.

endorama replied:
I set 30 seconds thinking that duration for a CI is in the order of minutes, so in the worst case is wait is 29 seconds in a multi minute run. What is your concern with this duration? I'm curious about it, in any case ok for me to reduce it. Would 10 seconds be ok?

marekdedic replied:
Exponential backoff with a maximum?

samcoe replied:
@endorama My thought was that when trying out the new functionality 30 seconds seemed like a long time to see refreshed results. CI runs do not have to be lengthy so a shorter refresh interval would work well for both shorter runs as well as longer ones. I think 10 seconds is a good starting point, users can always customize it to their needs with the flag.

endorama replied:
I'll change it to 10 seconds, but I like the idea of @marekdedic about the exponential backoff (with 1 minute maximum maybe?). Is it worth the gain in terms of complexity/user experience?

samcoe replied:
I don't think the trade off of complexity is worth it. We can always iterate on refresh interval later as we receive user feedback.

endorama replied:
Changed in 925c825

samcoe commented on pkg/cmd/pr/checks/checks.go:
```diff
@@ -59,22 +62,29 @@ func NewCmdChecks(f *cmdutil.Factory, runF func(*ChecksOptions) error) *cobra.Co
 	}
 
 	cmd.Flags().BoolVarP(&opts.WebMode, "web", "w", false, "Open the web browser to show details about checks")
+	cmd.Flags().BoolVarP(&opts.Watch, "watch", "", false, "Watch checks and wait until are finished running")
```
```suggestion
	cmd.Flags().BoolVarP(&opts.Watch, "watch", "", false, "Watch checks until they finish")
```

endorama replied:
Added in 1401172

samcoe commented on pkg/cmd/pr/checks/checks.go:
```diff
@@ -59,22 +62,29 @@ func NewCmdChecks(f *cmdutil.Factory, runF func(*ChecksOptions) error) *cobra.Co
 
 	cmd.Flags().BoolVarP(&opts.WebMode, "web", "w", false, "Open the web browser to show details about checks")
+	cmd.Flags().BoolVarP(&opts.Watch, "watch", "", false, "Watch checks and wait until are finished running")
+	cmd.Flags().IntVarP(&opts.Interval, "interval", "i", defaultInterval, "The number of seconds between each reprint When using --watch")
```
```suggestion
	cmd.Flags().IntVarP(&opts.Interval, "interval", "i", defaultInterval, "Refresh interval in seconds when using watch flag")
```

endorama replied:
Added in 1401172

samcoe commented on pkg/cmd/pr/checks/checks.go:
```diff
@@ -83,163 +93,75 @@ func checksRun(opts *ChecksOptions) error {
-	if len(rollup) == 0 {
-		return fmt.Errorf("no checks reported on the '%s' branch", pr.BaseRefName)
+func checksRunWatchMode(opts *ChecksOptions) error {
+	duration, err := time.ParseDuration(fmt.Sprintf("%ds", opts.Interval))
+	if err != nil {
+		return fmt.Errorf("could not parse interval: %w", err)
```
I think we should move the parsing of the interval flag up into the `RunE` portion of the command. This is a direction we have been moving lately trying to create a consistency through commands of when/where to parse and deal with flag logic. Once this code moves we will need to change the type of `interval` from `int` to `Duration`. Lastly, since this is a flag parsing error we can use the `cmdutil.FlagErrorf` helper function for returning the error.

endorama replied:
Changed in 6e30b62a

samcoe commented on pkg/cmd/pr/checks/checks.go:
```diff
@@ -83,163 +93,75 @@ func checksRun(opts *ChecksOptions) error {
 
-		elapsed := ""
-		zeroTime := time.Time{}
+	for meta.Pending > 0 {
```
What are your thoughts on turning this into a simple `for {` loop and doing the first retrieving and outputting of checks inside the loop and put the sleep at end of the loop instead of the top? This feels like it would make this code a bit simpler.

endorama replied:
I'm not sure I understood completely, but I tried a refactor changing as you suggested in f1b44b4a. Is this the direction you where thinking of?

samcoe commented on pkg/cmd/pr/checks/collect.go:
```diff
@@ -0,0 +1,94 @@
+	Skipping int
+}
+
+func collect(opts *ChecksOptions) ([]check, checksMeta, error) {
```
Right now it feels like this function is doing too much work. I think we should separate out the functionality of retrieving the checks from the API and aggregation of the checks into the `checks` and `meta` structs. Additionally, `collect` does not feel descriptive enough to me what do you think of `aggregateChecks` or `categorizeChecks`?

samcoe commented on pkg/cmd/pr/checks/output.go:
```diff
@@ -0,0 +1,108 @@
+	tp.EndRow()
+}
+
+func printTable(isTerminal bool, cs *iostreams.ColorScheme, meta checksMeta, checks []check, opts *ChecksOptions) error {
```
Similar to the `collect` function this feels like it is doing too much work. The output of the summary and table are two distinct parts of this function so that seems like a good place to break it up. What do you think?

samcoe reviewed (changes requested):
@endorama This is coming along nicely! I tested the functionality and everything worked well. My main concerns are around some of the complexity and code structure. I know some of it was already in place before you, but it makes sense to clean it up while we are working on it. Let me know if you have any questions about my comments.

endorama committed:
update flags help text

endorama committed:
update default interval to 10s

endorama committed:
move interval flag parse to RunE

endorama committed:
refactor checksRunWatchMode to use infinite loop

endorama commented:
@samcoe I did another round. I would wait for your review before moving on to refactoring `collect` and `printTable` as the direction for the `for {` wasn't perfectly clear to me.

samcoe committed:
Refactor printTable function

samcoe committed:
Refactor collect function

samcoe committed:
Set up checksRun to use new refactored functions and simplify logic a bit

samcoe committed:
Add tests

samcoe committed:
Always set interval in opts

samcoe reviewed (approved):
@endorama Thanks so much for the hard work and getting us 90% of the way done. We have a release coming up and in an effort to get this completed in time I went ahead and pushed some changes completing the refactors of `collect` and `printTable` that I mentioned in previous comments. Appreciate the help getting this feature completed 🙇

samcoe commented on pkg/cmd/pr/checks/aggregate.go:
```diff
@@ -0,0 +1,110 @@
+
+	rollup := pr.StatusCheckRollup.Nodes[0].Commit.StatusCheckRollup.Contexts.Nodes
+	if len(rollup) == 0 {
+		return checks, counts, fmt.Errorf("no checks reported on the '%s' branch", pr.HeadRefName)
```
@vilmibm I changed this from `pr.BaseRefName` to `pr.HeadRefName`, let me know if that makes sense. This is the only functionality outside of `--watch` mode that was changed.

vilmibm replied:
sure sounds more correct >_>

samcoe commented on pkg/cmd/pr/checks/checks.go:
```diff
@@ -65,208 +78,77 @@ func NewCmdChecks(f *cmdutil.Factory, runF func(*ChecksOptions) error) *cobra.Co
+		if err := opts.IO.EnableVirtualTerminalProcessing(); err == nil {
+			// ANSI escape code to clean terminal window
+			// https://stackoverflow.com/a/22892171
+			fmt.Print(opts.IO.Out, "\033[H\033[2J")
```
@vilmibm Can I get a double check on if we should be using this code to clear the screen or the one you used for `run watch`.

vilmibm replied:
we should be copying what I did for run watch (including the platform negotiation).

samcoe commented on pkg/cmd/pr/checks/checks.go:
```diff
@@ -65,208 +78,77 @@ func NewCmdChecks(f *cmdutil.Factory, runF func(*ChecksOptions) error) *cobra.Co
 
 	cmd.Flags().BoolVarP(&opts.WebMode, "web", "w", false, "Open the web browser to show details about checks")
+	cmd.Flags().BoolVarP(&opts.Watch, "watch", "", false, "Watch checks until they finish")
+	cmd.Flags().IntVarP(&interval, "interval", "i", defaultInterval, "Refresh interval in seconds when using `--watch` flag")
```
Cobra has a duration flag which I thought about using here but was unsure since we went with an int flag for `run watch`.

endorama replied:
My bad, I didn't know about the flag type. Nice catch! :+1:

endorama replied:
I went forward using a duration flag here, let me know if I should remove this or if you would like me to submit another PR to update `run watch` too.

endorama committed:
use Duration flag

vilmibm reviewed (changes requested):
Unless I'm missing a reason not to, this should use the same screen redrawing logic as `run watch`.

samcoe committed:
Revert back to using int flag for consistency with run watch

samcoe committed:
Use run watch screen clearing mechanism

samcoe committed:
Re-add pager support

samcoe committed:
Merge branch 'trunk' into pr-checks-wait

samcoe commented:
@vilmibm I made the changes requested, can you please take another look 🙇

vilmibm reviewed (approved):
thanks!!

endorama commented:
Thank you all!

Jasha10 commented:
The [release notes for gh v2.5.1](https://github.com/cli/cli/releases/tag/v2.5.1) incorrectly say that a `--wait` flag has been added, whereas the flag is now named `--watch`.
Also this PR should probably be renamed to mention `--watch` instead of `--wait`.
